aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1387 @@
+     b) Guide handling when another transaction observes a dependency during execution
+   - Helps constrain optimistic concurrency by limiting cascading aborts
+
+2. Behavior:
+   - Best-effort approach that allows flexibility in concurrency scenarios, but such that
+     high-priority transactions may still be re-executed even in stalled state
+**/
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) enum StatusEnum {
+    PendingScheduling,
+    Executing,
+    Aborted,
+    Executed,
+}
+
+/// Represents the internal execution status of a transaction at a specific incarnation.
+/// Tracks both the current state (via StatusEnum) and the incarnation number.
+/// Incarnation number, starting at 0 and incremented after each abort, represents a
+/// distinct execution attempt of the transaction.
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct InnerStatus {
+    status: StatusEnum,
+    incarnation: Incarnation,
+}
+
+impl InnerStatus {
+    fn new() -> Self {
+        Self {
+            status: StatusEnum::PendingScheduling,
+            incarnation: 0,
+        }
+    }
+
+    #[cfg(test)]
+    pub(crate) fn new_for_test(status: StatusEnum, incarnation: Incarnation) -> Self {
+        Self {
+            status,
+            incarnation,
+        }
+    }
+
+    fn try_start_executing(&mut self) -> Option<Incarnation> {
+        if self.status == StatusEnum::PendingScheduling {
+            self.status = StatusEnum::Executing;
+            return Some(self.incarnation);
+        }
+        None
+    }
+
+    fn incarnation(&self) -> Incarnation {
+        self.incarnation
+    }
+
+    fn never_started_execution(&self, incarnation: Incarnation) -> bool {
+        self.incarnation < incarnation
+            || (self.incarnation == incarnation && self.status == StatusEnum::PendingScheduling)
+    }
+
+    fn already_aborted(&self, incarnation: Incarnation) -> bool {
+        self.incarnation > incarnation
+            || (self.incarnation == incarnation && self.status == StatusEnum::Aborted)
+    }
+
+    fn pending_scheduling(&self) -> Option<Incarnation> {
+        (self.status == StatusEnum::PendingScheduling).then_some(self.incarnation)
+    }
+
+    fn is_executed(&self) -> bool {
+        self.status == StatusEnum::Executed
+    }
+
+    fn ever_executed(&self) -> bool {
+        use StatusEnum::*;
+        match self.status {
+            PendingScheduling | Executing | Aborted => self.incarnation > 0,
+            Executed => true,
+        }
+    }
+}
+
+/// Flag values for dependency resolution stored in an `AtomicU8` to allow lock-free reads.
+/// These values represent the state of a transaction that other transactions depend on.
+/// The flags are updated while holding the status lock but provide a fast way to evaluate
+/// a predicate associated with the status that enables the scheduler to make decisions about
+/// stall propagation, transaction scheduling, and dependency resolution.
+mod dependency_flags {
```

Comment:
Is it necessary to maintain this additional flag? Can't it be inferred from `StatusEnum`, maybe with some additional information about stalling?

Reply:
It can, main benefit is lock free reads on dependency flag because it's stored as an atomic U8.